OpenRA.Mods.RA2/Traits/Garrisoner.cs

```diff
@@ -0,0 +1,191 @@
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Drawing;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Mods.RA2.Activities;
+using OpenRA.Mods.RA2.Orders;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.RA2.Traits
+{
+    public enum AlternateTransportsMode { None, Force, Default, Always }
+
+    [Desc("This actor can enter Garrison actors.")]
+    public class GarrisonerInfo : ITraitInfo
+    {
+        public readonly string GarrisonType = null;
+        public readonly PipType PipType = PipType.Green;
+        public readonly int Weight = 1;
+
+        [Desc("Use to set when to use alternate transports (Never, Force, Default, Always).",
+            "Force - use force move modifier (Alt) to enable.",
+            "Default - use force move modifier (Alt) to disable.")]
+        public readonly AlternateTransportsMode AlternateTransportsMode = AlternateTransportsMode.Default;
+
+        [Desc("Number of retries using alternate transports.")]
+        public readonly int MaxAlternateTransportAttempts = 1;
+
+        [Desc("Range from self for looking for an alternate transport (default: 5.5 cells).")]
+        public readonly WDist AlternateTransportScanRange = WDist.FromCells(11) / 2;
+
+        [Desc("Valid Stances. Default Ally")]
+        public readonly Stance TargetStances = Stance.Ally;
+
+        [GrantedConditionReference]
+        [Desc("The condition to grant to when this actor is loaded inside any transport.")]
+        public readonly string GarrisonCondition = null;
+
+        [Desc("Conditions to grant when this actor is loaded inside specified transport.",
+            "A dictionary of [actor id]: [condition].")]
+        public readonly Dictionary<string, string> GarrisonConditions = new Dictionary<string, string>();
+
+        [GrantedConditionReference]
+        public IEnumerable<string> LinterGarrisonConditions { get { return GarrisonConditions.Values; } }
+
+        [VoiceReference] public readonly string Voice = "Action";
+
+        public object Create(ActorInitializer init) { return new Garrisoner(this); }
+    }
+
+    public class Garrisoner : INotifyCreated, IIssueOrder, IResolveOrder, IOrderVoice, INotifyRemovedFromWorld, INotifyEnteredGarrison, INotifyExitedGarrison
+    {
+        public readonly GarrisonerInfo Info;
+        public Actor Transport;
+
+        ConditionManager conditionManager;
+        int anyGarrisonToken = ConditionManager.InvalidConditionToken;
+        int specificGarrisonToken = ConditionManager.InvalidConditionToken;
+
+        public Garrisoner(GarrisonerInfo info)
+        {
+            Info = info;
+            Func<Actor, Actor, bool> canTarget = IsCorrectGarrisonType;
+            Func<Actor, Actor, bool> useEnterCursor = CanEnter;
+            Orders = new EnterGarrisonTargeter<GarrisonInfo>[]
+            {
+                new EnterGarrisonTargeter<GarrisonInfo>("EnterTransport", 5, canTarget, useEnterCursor, Info.AlternateTransportsMode)
```

Comment:
Is there a reason why are you reusing EnterTransport as the orderstring? Is this intentional to prevent a buggy behaviour or just lack of change yet? If the latter, please change it to EnterGarrison for further sandboxing.

Reply:
I fixed this in the current version.